src/index.js

```diff
@@ -251,7 +264,24 @@ function graphqlHTTP(options: Options): Middleware {
         }
 
         // Validate AST, reporting any errors.
-        const validationErrors = validate(schema, documentAST, validationRules);
+        const string =
+          JSON.stringify(schema) +
+          JSON.stringify(source) +
+          validationRules.toString();
```

Comment:
There is a problem with your caching approach `JSON.stringify(schema)` only stringify type names and not things like fields, enum values, etc. It also doesn't take into account user-provided callbacks that affect validation e.g. `parseLiteral`
https://github.com/graphql/graphql-js/blob/8c96dc8276f2de27b8af9ffbd71a4597d483523f/src/validation/rules/ValuesOfCorrectType.js#L186

Hashing `validationRules` is affected by the exact same problems.

I think that it's impossible to implement secure & effective validation caching for all use cases.
For example, in most cases, you schema and validation rules stay the same for all requests so converting them to string and hashing is an overhead.

I think the simplest and the most flexible solution would be to allow wrapping/replacing `validate` function, similar to #391 proposal but for `validate`.

@jamesmoriarty What do you think?

Reply:
I think you're right @IvanGoncharov. We were concerned we had unsupported edge cases. We're only using a subset of graphql's feature so nothing was obviously broken. I'll see if I can invert the cache implementation with the wrapper function.